File: examples/lua/masterslave/master.lua
-- Copyright (c) 2011-2012, 2014. The SimGrid Team.
-- All rights reserved.

-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the license (GNU LGPL) which comes with this package.

function Master(...)

  if select("#", ...) ~= 4 then
    error("Wrong number of arguments (got " .. #arg ..
        ", expected 4: nb_tasks comp_size comm_size slave_count)")
  end

  local nb_task, comp_size, comm_size, slave_count = select(1, ...)
  simgrid.info("Hello from lua, I'm the master")


  -- Dispatch the tasks

  for i = 1, nb_task do
    local task = simgrid.task.new("Task " .. i, comp_size, comm_size)
    local task_name = task:get_name()
    local alias = "slave " .. string.format("%d", i % slave_count)
    simgrid.info("Sending  '" .. task_name .. "' to '" .. alias .."'")
    task:send(alias) -- C user data set to NULL
    simgrid.info("Done sending '".. task_name .. "' to '" .. alias .."'")
  end

  -- Sending Finalize Message To Others

  simgrid.info("All tasks have been dispatched. Let's tell everybody the computation is over.")
  for i = 0, slave_count - 1 do
    local alias = "slave " .. i
    simgrid.info("Sending finalize to '" .. alias .. "'")
    local finalize = simgrid.task.new("finalize", comp_size, comm_size)
    finalize:send(alias)
  end
  simgrid.info("Everything's done.")
end -- end_of_master

File: examples/scala/masterslave/Master.scala
/*
 * Master of a basic master/slave example in Java
 *
 * Copyright (c) 2006-2013. The SimGrid Team.
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. 
 */
package masterslave;

import org.simgrid.msg.Host;
import org.simgrid.msg.Msg;
import org.simgrid.msg.MsgException;
import org.simgrid.msg.Task;
import org.simgrid.msg.Process;;

class Master(host:Host, name:String, args:Array[String]) extends Process(host,name,args) {
  def main(args:Array[String]) {
    if (args.length < 4) {
      Msg.info("Master needs 4 arguments")
      System.exit(1)
    }

    val tasksCount = args(0).toInt		
    val taskComputeSize = args(1).toDouble		
    val taskCommunicateSize = args(2).toDouble
    val slavesCount = args(3).toInt

    Msg.info("Hello! Got " +  slavesCount + " slaves and " + tasksCount + " tasks to process")

    for (i <- 0 until tasksCount) {
      val task = new Task("Task_" + i, taskComputeSize, taskCommunicateSize)
      task.send("slave_"+(i%slavesCount))
    }

    Msg.info("All tasks have been dispatched. Let's tell everybody the computation is over.")

    for (i <- 0 until slavesCount) {
      val task = new FinalizeTask()
      task.send("slave_"+(i%slavesCount))
    }

    Msg.info("Goodbye now!")
  }
}
